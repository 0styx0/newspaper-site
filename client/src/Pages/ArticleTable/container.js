import React from 'react';
import fetchFromApi from '../../helpers/fetchFromApi';

import ArticleTable from './';

export default class ArticleTableContainer extends React.Component {

    constructor() {
        super();

        this.putData = this.putData.bind(this);

        this.state = {
            history: []
        };
    }

    async putData(num = null) {

        const lastPath = +window.location.pathname.split("/").slice(-1);

        num = (isNaN(lastPath) || ((!isNaN(num)) && num)) ? num : lastPath;


        // use cached results (don't load new info)
        if (this.state.history[num]) {
            this.setState({
                articles: this.state.history[num].articles,
                issueInfo: this.state.history[num].issueInfo
            });

            window.history.pushState({}, `Issue ${num}`, num);
        }
        else {

            const data = await this.getData(num);
            const json = await data.json();

            window.history.pushState({}, `Issue ${num}`, isNaN(lastPath) ? "modifyArticles/"+json[2].num : json[2].num);

            this.setArticleInfo(json);
        }
    }

    componentWillMount() {

        this.putData(null);
    }

    async getData(num) {

         return await fetchFromApi(`articleGroup?articlesFor=${num}`);

    }


    setArticleInfo(data) {

        const history = [...this.state.history];
        history[data[2].num] = {articles: data[0], issueInfo: data[2]};

        this.setState({
            articles: data[0].length ? data[0] :  null,
            issueInfo: data[2],
            history
        });
    }

    render() {

            return <ArticleTable
                     articles={this.state.articles}
                     issue={this.state.issueInfo}
                     update={(e) => this.putData(e.target.value)}
                   />
    }
}
