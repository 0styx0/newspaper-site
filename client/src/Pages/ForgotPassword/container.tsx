import * as React from 'react';
import { graphql, withApollo } from 'react-apollo';
import { RecoverPassword } from '../../graphql/user';
import ForgotPassword from './';

interface Props {
    recoverPassword:
      (params: {
          query: typeof RecoverPassword,
          variables: { username: string, authCode: string, email: string }
        }
      ) => Promise<{message: string}>;
}

export class ForgotPasswordContainer extends React.Component<Props, {}> {

    constructor(props: Props) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);

    }

    /**
     * Sends data to server
     */
    onSubmit(target: HTMLFormElement) {

        this.props.recoverPassword({
            query: RecoverPassword,
            variables: {
                authCode: (target.querySelector('input[name=lastAuth]') as HTMLInputElement).value,
                username: (target.querySelector('input[name=username]') as HTMLInputElement).value,
                email: (target.querySelector('input[name=email]') as HTMLInputElement).value,
            }
        });
    }

    render() {
        return <ForgotPassword onSubmit={this.onSubmit} />;
    }
}

const ForgotPasswordContainerWithData =
    graphql(RecoverPassword, {name: 'recoverPassword'})(ForgotPasswordContainer as any);

export default withApollo(ForgotPasswordContainerWithData);
