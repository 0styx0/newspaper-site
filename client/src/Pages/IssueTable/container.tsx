import * as React from 'react';
import { compose, graphql } from 'react-apollo';
import { IssueQuery, IssueUpdate } from '../../graphql/issues';
import { Issue } from './interface.shared';
import IssueTable from './';

interface State {
    privateIssue: { // admins can change these (until public is true)
        public?: boolean;
        name?: string;
    };
    loaded: boolean;
}

interface Props {
    data: {
        loading: boolean;
        issues: Issue[]
    };
    mutate: Function;
}

/**
 * Creates a table of issues, in descending order with info of
 * number, name, datePublished, and views (sum of views of all articles in the issue)
 *
 * If user is admin (lvl 3) and issue is not public, user can change issue name and make it public.
 * In that case, @see IssueTable.allowEditsOfLastIssue for difference in UI
 */
export class IssueTableContainer extends React.Component<Props, State> {

    constructor() {
        super();

        this.onChangeIssueInfo = this.onChangeIssueInfo.bind(this);
        this.onSubmit = this.onSubmit.bind(this);

        this.state = {
            privateIssue: {},
            loaded: false
        };
    }

    /**
     * When graphql gets data from @see graphql/issues,
     * this turns it into 2d array of what will be table rows and saves it to state.issueInfo
     *
     * If current user is an admin, enables editing of most recent, unpublished issue @see this.allowEditsOfLastIssue
     */
    componentWillReceiveProps(props: Props) {

        if (!props.data.issues || this.props.data.issues.length > 0) {
            return;
        }

        this.setState({loaded: true});
    }

    /**
     * @param e {HTMLInputElement event}
     *
     * Adds input's value to this.state.privateIssue[input.name]
     *
     * @example if event.target = <input name="public" value="1" />, then after this, this.state.privateIssue.public = 1
     */
    onChangeIssueInfo(e: Event) {

        const target = e.target as HTMLInputElement;

        const privateIssue = Object.assign({}, this.state.privateIssue);
        privateIssue[target.name] = target.value;

        this.setState({
            privateIssue
        });
    }

    /**
     * @param e {HTMLFormElement event}
     *
     * Sends this.state.privateIssue data to server (so name and/or public status can be saved to db)
     */
    onSubmit(e: Event) {

        e.stopPropagation();
        e.preventDefault();

        this.props.mutate({
            variables: {
                name: this.state.privateIssue.name,
                public: !!this.state.privateIssue.public
            }
        });
    }

    render() {

        if (this.props.data.issues.length < 1) {
            return null;
        }

        return (
            <IssueTable
              key={+this.state.loaded}
              onSubmit={this.onSubmit}
              onChangeIssueInfo={this.onChangeIssueInfo}
              issues={this.props.data.issues}
            />
        );
    }
}


const IssueTableContainerWithData = compose(
    graphql(IssueQuery),
    graphql(IssueUpdate)
)(IssueTable as any);

export default IssueTableContainerWithData;