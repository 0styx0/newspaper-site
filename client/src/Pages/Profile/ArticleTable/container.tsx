import * as React from 'react';
import { graphql, withApollo, compose } from 'react-apollo';
import { ArticleDelete } from '../../../graphql/articles';
import { Article } from '../shared.interfaces';
import UserArticleTable from './';

interface Props {
    articles: Article[];
    deleteArticle: Function;
}

/**
 * Container for @see ./index.tsx
 */
export class UserArticleTableContainer extends React.Component<Props, {idsToDelete: Set<string>}> {

    constructor() {
        super();

        this.onDelete = this.onDelete.bind(this);
        this.onSubmit = this.onSubmit.bind(this);

        this.state = {
            idsToDelete: new Set<string>()
        };
    }

    /**
     * Adds an article to state.idsToDelete
     */
    onDelete(e: Event) {

        const idsToDelete: Set<string> = this.state.idsToDelete;
        const value = (e.target as HTMLInputElement).value;

        if (idsToDelete.has(value)) {

            idsToDelete.delete(value);
        } else {

            idsToDelete.add(value);
        }

        this.setState({
            idsToDelete
        });
    }

    /**
     * Sends idsToDelete to server to be deleted
     */
    onSubmit(e: Event) {

        e.preventDefault();
        e.stopPropagation();

        this.props.deleteArticle({
            variables: {
                ids: [...this.state.idsToDelete]
            }
        });
    }

    render() {

        return (
            <UserArticleTable
              articles={this.props.articles}
              onDelete={this.onDelete}
              onSubmit={this.onSubmit}
            />
        );
    }
}


const UserArticleTableContainerWithData = compose(
    graphql(ArticleDelete, {name: 'deleteArticle'})
)(UserArticleTableContainer as any);

export default withApollo(UserArticleTableContainerWithData);