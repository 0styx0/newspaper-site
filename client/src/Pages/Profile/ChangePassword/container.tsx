import * as React from 'react';
import ChangePassword from './';
import { UserUpdate } from '../../../graphql/user';
import { graphql, compose, withApollo } from 'react-apollo';

interface Props {
    updatePassword: Function;
}

export class ChangePasswordContainer extends React.Component<Props, {}> {

    constructor() {
        super();

        this.onSubmit = this.onSubmit.bind(this);
    }

    /**
     * Send data to server
     */
    onSubmit(e: Event) {

        const target = e.target as HTMLFormElement;

        const oldPassword = target.querySelector('[name=password]') as HTMLInputElement;
        const newPassword = target.querySelector('[name=newPassword]') as HTMLInputElement;
        const newPasswordConfirmation = target.querySelector('[name=newPasswordConfirmation]') as HTMLInputElement;

        if (newPassword.value !== newPasswordConfirmation.value ||
          !(oldPassword.value && newPassword.value && newPasswordConfirmation.value)) {

            return;
        }

        this.props.updatePassword({
            variables: {
                password: oldPassword.value,
                newPassword: newPassword.value
            }
        });
    }

    render() {
        return <ChangePassword onSubmit={this.onSubmit} />;
    }
}

const ChangePasswordContainerWithData = compose(
    graphql(UserUpdate, {name: 'updatePassword'}) as any
)(ChangePasswordContainer);

export default withApollo(ChangePasswordContainerWithData);