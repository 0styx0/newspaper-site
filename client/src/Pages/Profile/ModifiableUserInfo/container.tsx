import * as React from 'react';
import { PrivateUserQuery, UserUpdate } from '../../../graphql/user';
import { UserDelete } from '../../../graphql/users';
import { graphql, withApollo, compose } from 'react-apollo';
import { ModifiableUserInfo } from '../shared.interfaces';
import ModifiableUserInfoComponent from './';

interface Props {
    updateUser: Function;
    deleteUser: Function;
    privateUserData: {
        users: [ModifiableUserInfo] // always length = 1
    };
}

interface State {
    updates: {
        twoFactor?: boolean;
        notificationStatus?: boolean;
    };
    delete?: boolean;
}

/**
 * Container, has event listeners and passes data to @see ./index.tsx
 */
export class ModifiableUserInfoContainer extends React.Component<Props, State> {

    constructor() {
        super();

        this.onDelete = this.onDelete.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
        this.onChange = this.onChange.bind(this);

        this.state = {
            updates: {}
        };
    }

    onDelete(e: Event) {

        this.setState({
            delete: (e.target as HTMLInputElement).checked
        });
    }

    /**
     * Saves changes to state.updates
     *
     * @uses `e.target.name`, `e.target.value`
     */
    onChange(e: Event) {

        const target = e.target as HTMLInputElement;

        const updates = Object.assign({}, this.state.updates);

        updates[target.name] = target.checked;

        this.setState({
            updates
        });
    }

    /**
     * Sends `state.updates` to server
     */
    onSubmit(target: HTMLFormElement) {

        if (this.state.delete) {
            return this.deleteUser();
        }

        if (Object.keys(this.state.updates).length > 0) {
            this.props.updateUser({
                variables: Object.assign(this.state.updates, {
                    password: (target.querySelector('[name=password]') as HTMLInputElement).value
                })
            });
        }
    }

    /**
     * Sends id of user to server so user can be deleted
     */
    deleteUser() {

        this.props.deleteUser({
            variables: {
                ids: [this.props.privateUserData.users[0].id]
            }
        });
    }

    render() {

        if (!this.props.privateUserData || !this.props.privateUserData.users) {
           return null;
        }

        return (
            <ModifiableUserInfoComponent
              onSubmit={this.onSubmit}
              onChange={this.onChange}
              onDelete={this.onDelete}
              {...this.props.privateUserData.users[0]}
            />
        );
    }
}

const ModifiableUserInfoContainerWithData = compose(
    graphql(PrivateUserQuery, {
        name: 'privateUserData',
        options: {
            variables: {
                profileLink: 'meiselesd2018'
            }
        }
    }),
    graphql(UserUpdate, {name: 'updateUser'}),
    graphql(UserDelete, {name: 'deleteUser'})
)(ModifiableUserInfoContainer as any);

export default withApollo(ModifiableUserInfoContainerWithData);
