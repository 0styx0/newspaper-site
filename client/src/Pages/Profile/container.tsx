import * as React from 'react';
import { UserQuery } from '../../graphql/user';
import { graphql, compose, withApollo } from 'react-apollo';
import Profile from './';
import { Article, PublicUserInfo } from './shared.interfaces';

interface Props {
    data: {
        users: {
            articles: Article[];
        } & PublicUserInfo
    };
    client: {
        query: ( params: { query: typeof UserQuery, variables: { profileLink: string; } } ) => Promise<Props>;
    };
}

interface State {
    articles: Article[];
    user: PublicUserInfo;
}

class ProfileContainer extends React.Component<Props, State> {

    public state: State;

    constructor(props: Props) {
        super(props);

        this.state = {
            user: {} as PublicUserInfo,
            articles: [] as Article[]
        };

    }

    async componentWillMount() {

        const path = window.location.pathname.split('/');

        const { data } = await this.props.client.query({
            query: UserQuery,
            variables: {
                profileLink: path[path.length - 1]
            }
        });

        this.setState({
            user: data.users[0],
            articles: data.users[0].articles
        });
    }

    render() {

        if (!this.state.user) {
            return null;
        }

        const articles = this.state.articles as Article[];
        const user = this.state.user as PublicUserInfo;

        return (
            <Profile
                articles={articles}
                user={user}
            />
        );
    }
}

const ProfileContainerWithData = compose(
    graphql(UserQuery, {
        options: {
            variables: {
                profileLink: window.location.pathname.split('/')[window.location.pathname.split('/').length - 1]
            }
        }
    })
)(ProfileContainer);

export default withApollo(ProfileContainerWithData as any) as any;
