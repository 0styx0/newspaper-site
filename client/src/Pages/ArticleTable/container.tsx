import * as React from 'react';
import { ArticleQuery, ArticleUpdate, ArticleDelete } from '../../graphql/articles';
import { compose, graphql, withApollo } from 'react-apollo';

import ArticleTable from './';

export interface Article {
    tags: {
            all: string[];
        };
    url: string;
    id: string;
    displayOrder: number;
    dateCreated: string;
    views: number;
    author: {
        fullName: string;
        profileLink: string;
    };
}

export interface Issue {
    num: number;
    max: number;
}

interface Props {
    data: {
        loading: boolean;
        issues?: [
            Issue & {
                articles: Article[]
            }
        ]; // will never be more length than 1
    };
    client: {
        query: ( params: { query: typeof ArticleQuery, variables: { issue: number | null; } } ) => Promise<Props>;
    };
    updateArticle: Function;
    deleteArticle: Function;
}

interface State {
    issue: Issue;
    articles: Article[];
    updates: {
        idsToDelete: Set<string>;
        displayOrder: Map<string, number>; // id => info
        tags: Map<string, string[]>;
    };
}

/**
 * Controller of ArticleTable (@see ./index)
 *
 * Gets data from server, then renders ArticleTable with it.
 * This holds listeners for all events that @see ArticleTable can fire
 */
export class ArticleTableContainer extends React.Component<Props, State> {

    constructor() {
        super();

        this.putData = this.putData.bind(this);
        this.convertPropsToState = this.convertPropsToState.bind(this);
        this.onChange = this.onChange.bind(this);
        this.onDelete = this.onDelete.bind(this);
        this.onSubmit = this.onSubmit.bind(this);



        this.state = {
            issue: {} as Issue,
            articles: [],
            updates: {
                idsToDelete: new Set<string>(),
                displayOrder: new Map<string, number>(),
                tags: new Map<string, string[]>()
            }
        };
    }

    /**
     * @param num - articles from what issue to get. If null, get latest
     *
     * Gets data needed for table (@see State.Article interface) and changes page url to modifyArticles/currentIssue
     *
     * Depends on window.location.pathname
     */
    async putData(num: number | null = null) {

        const lastPath = +window.location.pathname.split('/').slice(-1)[0];

        num = (isNaN(lastPath) || ((!isNaN(+num!)) && num)) ? num : lastPath;

        this.props.client.query({
            query: ArticleQuery,
            variables: {
                issue: num
            }
        }).then((data: Props) => {
            this.convertPropsToState(data);
        });

        window.history.pushState(
            {},
            `Issue ${num}`, isNaN(lastPath) ?
              `modifyArticles/${num}` :
             num + ''
        );
    }

    /**
     * Seeds state with graphql props
     */
    componentWillReceiveProps(newProps: Props) {
        this.convertPropsToState(newProps);
    }

    /**
     * Sets `state.issue` and `state.articles` to values extracted from props
     */
    convertPropsToState(props: Props) {

        if (!props.data.issues) {
            return;
        }

        let {max, num} = props.data.issues![0];

        this.setState({
            issue: {
                max,
                num
            },
            articles: props.data.issues![0].articles
        });
    }

    /**
     * Saves changes to state so can be submitted later on. This is for updates only. @see onDelete for deletions
     *
     * @uses `e.target.name` as index of state.updates
     * @uses `e.target.value` as the value
     */
    onChange(e: Event, article: Article) {

        e.stopPropagation();
        e.preventDefault();

        const target = e.target as HTMLInputElement | HTMLSelectElement;

        const stateUpdate = this.state.updates;

        const modifyMapCopy = new Map(stateUpdate[target.name]);
        let value: number | string[];

        if ((target as HTMLSelectElement).selectedOptions) { // for tags

            value = Array.from((target as HTMLSelectElement).selectedOptions)
                    .map(option => option.value)
                    .slice(-3)
                    .sort(); // using sort to make it easier to test, no actual reason besides that
        } else {
            value = +target.value;
        }

        modifyMapCopy.set(article.id, value);

        stateUpdate[target.name] = modifyMapCopy;

        this.setState({
            updates: stateUpdate
        });
    }

    /**
     * Adds/removes user's id from list of users to delete
     *
     * @uses e.target.checked as way to see if should add or remove e.target.value from list
     *
     * @param e {HTMLCheckboxEvent}
     */
    onDelete(e: Event) {

        const stateUpdate = this.state.updates;

        const target = e.target as HTMLInputElement;

        target.checked ? stateUpdate.idsToDelete.add(target.value) : stateUpdate.idsToDelete.delete(target.value);

        this.setState({
            updates: stateUpdate
        });
    }

    /**
     * Sends data from `state.updates` to server
     */
    onSubmit(e: Event) {

        e.stopPropagation();
        e.preventDefault();

        // separating update and delete into functions just to show they're separate stuff
        const submitUpdated = () => {

            const { displayOrder, tags } = this.state.updates;

            if (displayOrder.size + tags.size < 1) {
                return;
            }

            let data: {id: string, displayOrder?: number, tags?: string[]}[] = [...displayOrder].map(mapping => ({

                id: mapping[0],
                displayOrder: mapping[1]
            }));

            [...tags].forEach(mapping => {

                const idIdx = data.findIndex(elt => elt.id === mapping[0]);

                if (idIdx !== -1) {
                    data[idIdx].tags = mapping[1];
                } else {
                    data.push({
                        id: mapping[0],
                        tags: mapping[1]
                    });
                }
            });

            this.props.updateArticle({
                variables: {
                    data
                }
            });
        };

        const submitDeleted = () => {

            if (this.state.updates.idsToDelete.size > 0) {

                this.props.deleteArticle({
                    variables: {
                        ids: [...this.state.updates.idsToDelete]
                    }
                });
            }
        };

        submitUpdated();
        submitDeleted();
    }

    render() {

        if (this.state.articles.length < 1) {
            return null;
        }

        return (
            <ArticleTable
                articles={this.state.articles}
                key={this.state.issue.num}
                issue={this.state.issue}
                onChange={this.onChange}
                onDelete={this.onDelete}
                onUpdate={(e: Event) => this.putData(+(e.target as HTMLInputElement).value)}
                onSubmit={this.onSubmit}
            />
        );
    }
}

const ArticleTableContainerWithData = compose(
    graphql(ArticleQuery, {
        options: {
            variables: {
                issue: +window.location.pathname.split('/').slice(-1)[0]
            }
        }
    }),
    graphql(ArticleUpdate, {name: 'updateArticle'}),
    graphql(ArticleDelete, {name: 'deleteArticle'})
)(ArticleTableContainer as any);

export default withApollo(ArticleTableContainerWithData);