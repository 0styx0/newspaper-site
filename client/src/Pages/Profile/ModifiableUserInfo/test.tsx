import * as React from 'react';
import { ModifiableUserInfoContainer, Props, State } from './container';
import * as renderer from 'react-test-renderer';
import * as casual from 'casual';
import { randomCheckboxToggle, setInput, submitForm } from '../../../tests/enzyme.helpers';
import * as sinon from 'sinon';
import { mount, ReactWrapper } from 'enzyme';
import { ModifiableUserInfo } from '../shared.interfaces';
import userData from './__snapshots__/user.examples';

casual.define('user', (twoFactor: boolean, notifications: boolean): ModifiableUserInfo => {

    return {
        email: casual.email,
        twoFactor,
        notifications,
        id: casual.word
    };
});

type CustomCasualData = typeof casual & {user: (twoFactor: boolean, notifications: boolean) => ModifiableUserInfo};

const customCasual = casual as CustomCasualData;

function setup(
    userOptions: {notifications: boolean, twoFactor: boolean} = { notifications: false, twoFactor: false },
    mockGraphql: { updateUser?: Function, deleteUser?: Function } = {}
) {

    const mockFunc = async () => true;

    const wrapper = mount(
        <ModifiableUserInfoContainer
            updateUser={mockGraphql.updateUser || mockFunc}
            deleteUser={mockGraphql.deleteUser || mockFunc}
            fetchPrivateUserData={{
                refetch: async () =>
                    ({ data: { users: [customCasual.user(userOptions.twoFactor, userOptions.notifications)] } })
            }}
        />
    );

    wrapper.setState({
        privateUserData: {
            users: [customCasual.user(userOptions.twoFactor, userOptions.notifications)]
        }
    });

    return wrapper;
}

describe('<ModifiableUserInfoContainer>', () => {

    describe('snapshots', () => {

        function testSnap(data: ModifiableUserInfo) {

            const component = (
                <ModifiableUserInfoContainer
                    updateUser={() => true}
                    deleteUser={() => true}
                    fetchPrivateUserData={{ refetch: () => ({ data: { users: [data] } }) }}
                />
            );

            const tree = renderer.create(component);

            expect(tree.toJSON()).toMatchSnapshot();
        }

        it('should have notifications checked if notifications is true',
           () => testSnap(userData.onlyNotifications));

        it('should have twoFactor checked if twoFactor is true', () => testSnap(userData.only2fa));

        it('should have both checked if both are true', () => testSnap(userData.bothTrue));

        it('should have none checked if both are false', () => testSnap(userData.bothFalse));
    });

    describe('modifiable inputs', () => {

        /**
         * Tests if input[name=<name>] add/removes from state.updates.<name> when expected
         *
         * @param name - of input to check
         * @param pathToState - index of `state` to check
         */
        function toggleChecker(name: string, pathToState: string[] = ['updates', name]) {

            /**
             * Toggles checkbox and sees if state was changed
             *
             * @param shouldCheck - if final state of checkbox will be checked
             */
            function toggleAndCheckState(shouldCheck: boolean) {

                const userOptions = {notifications: false, twoFactor: false};
                userOptions[name] = !shouldCheck; // make sure that initial state is not what is expected

                const wrapper = setup(userOptions);

                randomCheckboxToggle(wrapper.find(`input[name="${name}"]`)); // not random in this case since only 1

                const plainState = (wrapper.find(ModifiableUserInfoContainer)
                    .instance() as {} as ModifiableUserInfoContainer).state;

                // adds pathToState to `component.state`
                const stateToUpdate = pathToState.reduce((accum, elt) => accum[elt], plainState);

                if (shouldCheck) {
                    expect(stateToUpdate).toBeTruthy();
                } else {
                    expect(stateToUpdate).toBeFalsy();
                }
            }

            it(`gets added to \`state\` when toggled on`, () => toggleAndCheckState(true));

            it(`gets removed from \`state\` when toggled off`,
               () => toggleAndCheckState(false));
        }

        describe('twoFactor', () => toggleChecker('twoFactor'));

        describe('notifications', () => toggleChecker('notifications'));

        describe('input [name=delAcc]', () => {

            let wrapper: ReactWrapper<Props, State>;
            let deleteBox: ReactWrapper<HTMLInputElement, {}>; // don't know if this is correct type

            /**
             * Sets `wrapper`and `deleteBox` and toggles the deleteBox once
             */
            beforeEach(() => {

                wrapper = setup();
                deleteBox = wrapper
                    .find(`input[name="delAcc"]`) as {} as ReactWrapper<HTMLInputElement, {}>;

                randomCheckboxToggle(deleteBox); // not random in this case since only 1
            });

            it('makes `state.delete` true if toggled on', () => {

                const deleteState = (wrapper
                    .find(ModifiableUserInfoContainer)
                    .instance() as {} as ModifiableUserInfoContainer)
                    .state
                    .delete;

                expect(deleteState).toBeTruthy();
            });

            it('makes `state.delete` false if toggled off', () => {

                // toggle off
                randomCheckboxToggle(deleteBox);

                const deleteState = (wrapper
                    .find(ModifiableUserInfoContainer)
                    .instance() as {} as ModifiableUserInfoContainer)
                    .state
                    .delete;

                expect(deleteState).toBeFalsy();
            });
        });
    });

    describe('submitting data', () => {

        describe('should send same data as is in `state`', () => {

            describe('when only 1 has changed', () => {

                [
                    ['twoFactor', 'twoFactor'],
                    ['notifications', 'notifications'],
                ].forEach(elt =>

                    test(`${elt[0]} has changed`, () => {

                        const expectedValue = true;
                        let password = '';

                        const wrapper = setup(
                            {notifications: false, twoFactor: false},
                            {updateUser: async (data: {variables: {twoFactor?: boolean, notifications?: boolean}}) => {

                                const expected = {
                                    password
                                };
                                expected[elt[1]] = expectedValue;

                                expect(data.variables).toEqual(expected);
                        }});

                        (wrapper
                            .find(ModifiableUserInfoContainer)
                            .instance() as {} as ModifiableUserInfoContainer)
                            .state
                            .updates[elt[1]] = expectedValue;

                        password = setInput(wrapper);
                        submitForm(wrapper);
                    })
                );

                test('when deleteAcc has changed', () => {

                    const expectedValue: string[] = [];

                    const wrapper = setup(
                        { notifications: false, twoFactor: false },
                        {deleteUser: async (data: {variables: {ids: [string]}}) => {

                            expect(data.variables).toEqual({
                                ids: expectedValue
                            });
                    }});

                    expectedValue.push((wrapper
                        .find('input[name="delAcc"]')
                        .instance() as {} as HTMLInputElement)
                        .value);

                    (wrapper
                        .find(ModifiableUserInfoContainer)
                        .instance() as ModifiableUserInfoContainer)
                        .setState({
                            delete: true
                        });

                    submitForm(wrapper);
                });
            });
        });

        test('delete and an update has changed', () => {

            const expectedDeleteValue: string[] = [];
            const expectedUpdateValue = true;
            const fieldToUpdate = casual.coin_flip ? 'noticationStatus' : 'twoFactor';

            const wrapper = setup(
                {notifications: false, twoFactor: false},
                {
                    deleteUser: async (data: {variables: {ids: [string]}}) => {

                        expect(data.variables).toEqual({
                            id: expectedDeleteValue
                        });
                    },
                    updateUser: async (data: {variables: {twoFactor?: boolean, notifications?: boolean}}) => {

                        const expected = {};
                        expected[fieldToUpdate] = expectedUpdateValue;

                        expect(data.variables).toEqual(expected);
                    }
                }
            );

            expectedDeleteValue.push(
                (wrapper
                    .find('input[name="delAcc"]')
                    .instance() as {} as HTMLInputElement
                ).value
            );

            (wrapper
                .find(ModifiableUserInfoContainer)
                .instance() as ModifiableUserInfoContainer)
                .setState({
                    delete: true
                });

            (wrapper
                .find(ModifiableUserInfoContainer)
                .instance() as {} as ModifiableUserInfoContainer)
                .state.updates[fieldToUpdate] = expectedUpdateValue;
        });

        test('2 updates have changed', () => {

            const expectedValue = true;
            let password = '';

            const wrapper = setup(
                {notifications: false, twoFactor: false},
                {updateUser: async (data: {variables: {twoFactor?: boolean, notifications?: boolean}}) => {

                const expected = {
                    notifications: expectedValue,
                    twoFactor: expectedValue,
                    password
                };

                expect(data.variables).toEqual(expected);
            }});

            (wrapper
                .find(ModifiableUserInfoContainer)
                .instance() as {} as ModifiableUserInfoContainer)
                .setState({
                    updates: {
                        notifications: expectedValue,
                        twoFactor: expectedValue
                    }
                });

            password = setInput(wrapper);

            wrapper.find('form').first().simulate('submit');
        });

        it(`shouldn't send if no data`, () => {

            const spy = sinon.spy();

            const wrapper = setup(
                {notifications: false, twoFactor: false},
                {
                    updateUser: async () => spy(),
                    deleteUser: async () => spy()
                }
            );

            submitForm(wrapper);

            expect(spy.called).toBeFalsy();
        });
    });
});
