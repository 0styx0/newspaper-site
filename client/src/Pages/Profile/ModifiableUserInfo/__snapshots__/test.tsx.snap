// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModifiableUserInfoContainer> snapshots should have both checked if both are true 1`] = `
<section
  className="container tableContainer"
>
  <h1>
    Options
  </h1>
  <form
    onSubmit={[Function]}
  >
    <table>
      <thead>
        <tr>
          <th>
             
            Email
          </th>
          <th>
             
            2FA
          </th>
          <th>
             
            Notifications
          </th>
          <th>
             
            <span
              className="danger"
            >
              Delete Account
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Lockman_Quincy@gmail.com
          </td>
          <td>
            <input
              defaultChecked={true}
              name="twoFactor"
              onChange={[Function]}
              type="checkbox"
            />
          </td>
          <td>
            <input
              defaultChecked={true}
              name="notificationStatus"
              onChange={[Function]}
              type="checkbox"
            />
          </td>
          <td>
            <input
              name="delAcc"
              onChange={[Function]}
              type="checkbox"
              value="totam"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <label>
      Password: 
      <input
        name="password"
        required={true}
        type="password"
      />
      <span
        className="danger"
      >
        *
      </span>
      
    </label>
    <input
      type="submit"
    />
  </form>
</section>
`;

exports[`<ModifiableUserInfoContainer> snapshots should have none checked if both are false 1`] = `
<section
  className="container tableContainer"
>
  <h1>
    Options
  </h1>
  <form
    onSubmit={[Function]}
  >
    <table>
      <thead>
        <tr>
          <th>
             
            Email
          </th>
          <th>
             
            2FA
          </th>
          <th>
             
            Notifications
          </th>
          <th>
             
            <span
              className="danger"
            >
              Delete Account
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Fadel.Ephraim@gmail.com
          </td>
          <td>
            <input
              defaultChecked={false}
              name="twoFactor"
              onChange={[Function]}
              type="checkbox"
            />
          </td>
          <td>
            <input
              defaultChecked={false}
              name="notificationStatus"
              onChange={[Function]}
              type="checkbox"
            />
          </td>
          <td>
            <input
              name="delAcc"
              onChange={[Function]}
              type="checkbox"
              value="dolorem"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <label>
      Password: 
      <input
        name="password"
        required={true}
        type="password"
      />
      <span
        className="danger"
      >
        *
      </span>
      
    </label>
    <input
      type="submit"
    />
  </form>
</section>
`;

exports[`<ModifiableUserInfoContainer> snapshots should have notifications checked if notificationStatus is true 1`] = `
<section
  className="container tableContainer"
>
  <h1>
    Options
  </h1>
  <form
    onSubmit={[Function]}
  >
    <table>
      <thead>
        <tr>
          <th>
             
            Email
          </th>
          <th>
             
            2FA
          </th>
          <th>
             
            Notifications
          </th>
          <th>
             
            <span
              className="danger"
            >
              Delete Account
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Kassulke.Glen@yahoo.com
          </td>
          <td>
            <input
              defaultChecked={false}
              name="twoFactor"
              onChange={[Function]}
              type="checkbox"
            />
          </td>
          <td>
            <input
              defaultChecked={true}
              name="notificationStatus"
              onChange={[Function]}
              type="checkbox"
            />
          </td>
          <td>
            <input
              name="delAcc"
              onChange={[Function]}
              type="checkbox"
              value="dolor"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <label>
      Password: 
      <input
        name="password"
        required={true}
        type="password"
      />
      <span
        className="danger"
      >
        *
      </span>
      
    </label>
    <input
      type="submit"
    />
  </form>
</section>
`;

exports[`<ModifiableUserInfoContainer> snapshots should have twoFactor checked if twoFactor is true 1`] = `
<section
  className="container tableContainer"
>
  <h1>
    Options
  </h1>
  <form
    onSubmit={[Function]}
  >
    <table>
      <thead>
        <tr>
          <th>
             
            Email
          </th>
          <th>
             
            2FA
          </th>
          <th>
             
            Notifications
          </th>
          <th>
             
            <span
              className="danger"
            >
              Delete Account
            </span>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            Bailey.Elnora@Lehner.us
          </td>
          <td>
            <input
              defaultChecked={true}
              name="twoFactor"
              onChange={[Function]}
              type="checkbox"
            />
          </td>
          <td>
            <input
              defaultChecked={false}
              name="notificationStatus"
              onChange={[Function]}
              type="checkbox"
            />
          </td>
          <td>
            <input
              name="delAcc"
              onChange={[Function]}
              type="checkbox"
              value="optio"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <label>
      Password: 
      <input
        name="password"
        required={true}
        type="password"
      />
      <span
        className="danger"
      >
        *
      </span>
      
    </label>
    <input
      type="submit"
    />
  </form>
</section>
`;
